mesonbuild/interpreter.py

```diff
@@ -943,7 +943,11 @@ def current_build_dir_method(self, args, kwargs):
         return os.path.join(src, sub)
 
     def source_root_method(self, args, kwargs):
-        return self.interpreter.environment.source_dir
+        if self.interpreter.subproject == '':
+            return self.interpreter.environment.source_dir
+        return os.path.join(self.interpreter.environment.source_dir,
+                            self.interpreter.subproject_dir,
+                            self.interpreter.subproject)
 
     def build_root_method(self, args, kwargs):
```

Comment:
Absolutely. We should also set `MESON_SOURCE_ROOT` and `MESON_BUILD_ROOT` per-subproject somehow. Right now those are also wrong for subprojects afaict.